### ISSUE TITLE
UNSUPPORTED_PROTOCOL with python:3.7.4-slim-buster docker image

### ISSUE THREAD

**Issue by daveisfera** [PyMySQL/PyMySQL#817]
We updated our docker image from python:3.7-4-slim-stretch to python:3.7.4-slim-buster and this error started happening when connecting to a MariaDB and MySQL databases using `PyMySQL` as the driver for `SQLAlchemy`:

```
2003, "Can't connect to MySQL server on 'mysqlnumetric.ceoxewvigov8.us-east-2.rds.amazonaws.com' ([SSL: UNSUPPORTED_PROTOCOL] unsupported protocol (_ssl.c:1076))
```

https://github.com/sqlalchemy/sqlalchemy/issues/4937

**Comment by methane** [comment]
The error is TLS protocol mismatch, not MySQL protocol.

You did not provide any information about TLS settting of the server.
So I can not answer why you faced the error.

I suppose your DB server uses old SSL.  And your Python in new docker image doesn't allow it.
If so, you can create custom ssl context which allows old SSL protocol and pass it to PyMySQL.

**Comment by daveisfera** [comment]
Running with `mysqlclient` on both stretch and buster works without any issues or any special changes, so I believe that this is something specific to `PyMySQL`

**Comment by methane** [comment]
So what?
This is clearly SSL error and you don't provide any information about your server information.

There are some possible cause which make different behaviour between mysqlclient and PyMySQL.  But I can not say anything without detailed information about your environment.

**Comment by daveisfera** [comment]
It's a MySQL database in RDS using 5.6.35 and the only change we made was to switch the docker base image from `python:3.7.4-slim-stretch` to `python:3.7.4-slim-buster` and this error started happening. Is there any other information that I could provide?

**Comment by methane** [comment]
> Is there any other information that I could provide?

I wrote:

> If so, you can create custom ssl context which allows old SSL protocol and pass it to PyMySQL.

Have you tried this?  You can specify TLS version explicitly.  MySQL 5.6.35 RDS seems supporting only TLS.

And when you connecting from mysqlclient, you are really using TLS?
The libmysqlclient falls back to plain connection when TLS is not usable.  It is very bad behavior because it may use plain connection even when you believed you are using secure connection.

You can check it by `show session status like 'Ssl_%'`.

**Comment by daveisfera** [comment]
Thanks for you help and I definitely don't want to be dropping TLS. Could the issue with that an SSL related library that used to be installed with `stretch` isn't installed with `buster`. Is there a way that I can check that?

**Comment by methane** [comment]
> Could the issue with that an SSL related library that used to be installed with `stretch` isn't installed with `buster`.

I don't think so.  If SSL is unavailable, another error must happen.
In your case, UNSUPPORTED_PROTOCOL error has happened.  It means SSL library is available and the error is raised while SSL handshake.

I suspect SSL in buster doesn't allow TLS v1.0 by default but your RDS server only supports TLS v1.0.  You can enable TLS v1.0 explicitly.

**Comment by MatrixManAtYrService** [comment]
I've encountered this as well, it happens in these images:
 
 - python:3.7
 - maven:3.6.3-jdk-11-slim

But not in this one

- centos:centos7

Also, I think that this is related:
https://stackoverflow.com/questions/58268626/ssl-unsupported-protocol-python-3-in-docker

My server's ssl-like variables are:
```
"show session status like 'Ssl_%'\G"
*************************** 1. row ***************************
Variable_name: Ssl_accept_renegotiates
        Value: 0
*************************** 2. row ***************************
Variable_name: Ssl_accepts
        Value: 1907383
*************************** 3. row ***************************
Variable_name: Ssl_callback_cache_hits
        Value: 0
*************************** 4. row ***************************
Variable_name: Ssl_cipher
        Value: ECDHE-RSA-AES256-GCM-SHA384
*************************** 5. row ***************************
Variable_name: Ssl_cipher_list
        Value: ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-DSS-AES128-GCM-SHA256:DHE-RSA-AES128-SHA256:DHE-DSS-AES128-SHA256:DHE-DSS-AES256-GCM-SHA384:DHE-RSA-AES256-SHA256:DHE-DSS-AES256-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:DHE-DSS-AES128-SHA:DHE-RSA-AES128-SHA:DHE-RSA-AES256-SHA:AES128-GCM-SHA256:DH-DSS-AES128-GCM-SHA256:ECDH-ECDSA-AES128-GCM-SHA256:AES256-GCM-SHA384:DH-DSS-AES256-GCM-SHA384:ECDH-ECDSA-AES256-GCM-SHA384:AES128-SHA256:DH-DSS-AES128-SHA256:ECDH-ECDSA-AES128-SHA256:AES256-SHA256:DH-DSS-AES256-SHA256:ECDH-ECDSA-AES256-SHA384:AES128-SHA:DH-DSS-AES128-SHA:ECDH-ECDSA-AES128-SHA:AES256-SHA:DH-DSS-AES256-SHA:ECDH-ECDSA-AES256-SHA:DHE-RSA-AES256-GCM-SHA384:DH-RSA-AES128-GCM-SHA256:ECDH-RSA-AES128-GCM-SHA256:DH-RSA-AES256-GCM-SHA384:ECDH-RSA-AES256-GCM-SHA384:
*************************** 6. row ***************************
Variable_name: Ssl_client_connects
        Value: 0
*************************** 7. row ***************************
Variable_name: Ssl_connect_renegotiates
        Value: 0
*************************** 8. row ***************************
Variable_name: Ssl_ctx_verify_depth
        Value: 18446744073709551615
*************************** 9. row ***************************
Variable_name: Ssl_ctx_verify_mode
        Value: 5
*************************** 10. row ***************************
Variable_name: Ssl_default_timeout
        Value: 7200
*************************** 11. row ***************************
Variable_name: Ssl_finished_accepts
        Value: 1907375
*************************** 12. row ***************************
Variable_name: Ssl_finished_connects
        Value: 0
*************************** 13. row ***************************
Variable_name: Ssl_server_not_after
        Value: Jun 12 08:26:09 2021 GMT
*************************** 14. row ***************************
Variable_name: Ssl_server_not_before
        Value: Jun 15 08:26:09 2011 GMT
*************************** 15. row ***************************
Variable_name: Ssl_session_cache_hits
        Value: 0
*************************** 16. row ***************************
Variable_name: Ssl_session_cache_misses
        Value: 0
*************************** 17. row ***************************
Variable_name: Ssl_session_cache_mode
        Value: SERVER
*************************** 18. row ***************************
Variable_name: Ssl_session_cache_overflows
        Value: 1115
*************************** 19. row ***************************
Variable_name: Ssl_session_cache_size
        Value: 128
*************************** 20. row ***************************
Variable_name: Ssl_session_cache_timeouts
        Value: 0
*************************** 21. row ***************************
Variable_name: Ssl_sessions_reused
        Value: 0
*************************** 22. row ***************************
Variable_name: Ssl_used_session_cache_entries
        Value: 128
*************************** 23. row ***************************
Variable_name: Ssl_verify_depth
        Value: 18446744073709551615
*************************** 24. row ***************************
Variable_name: Ssl_verify_mode
        Value: 5
*************************** 25. row ***************************
Variable_name: Ssl_version
        Value: TLSv1.2
```

I'd happily gather more info about the server what causes this or interrogate the various images where it does and doesn't happen, but I'd need some guidance on what to look for.

**Comment by methane** [comment]
> I've encountered this as well

Oh, really?  The error you seen is UNSUPPORTED_PROTOCOL?

I thought this issue is caused by mismatch between new OpenSSL on Debian (don't allow old protocols) and old MySQL (doesn't support TLS v1.2).  So I adviced that configure the ssl to allow old protocol (and no reply from the issue author).

But in your case, you can connect to the server with TLS v1.2:

```
*************************** 25. row ***************************
Variable_name: Ssl_version
        Value: TLSv1.2
```

What is your exact server version?  How did you install it?
What is shown when `show status like "%tls%"`?

**Comment by MatrixManAtYrService** [comment]
I have two servers that show this behavior.  I don't know who configured them, but I might be able to hunt them down and get a question answered if I had a specific question to ask.  Neither of them show any results for `show status like "%tls"`.  Here's what they do show:
```
❯ # SERVER 1
mysql -u$MYSQLUSER -p$MYSQLPASS -h"$MYSQLHOST1" -e"show variables like '%version%';"
+-------------------------+----------------------------------------------------+
| Variable_name           | Value                                              |
+-------------------------+----------------------------------------------------+
| innodb_version          | 5.7.18-16                                          |
| protocol_version        | 10                                                 |
| slave_type_conversions  |                                                    |
| tls_version             | TLSv1,TLSv1.1,TLSv1.2                              |
| version                 | 5.7.18-16-log                                      |
| version_comment         | Percona Server (GPL), Release 16, Revision d7301f8 |
| version_compile_machine | x86_64                                             |
| version_compile_os      | Linux                                              |
+-------------------------+----------------------------------------------------+
mysql -u$MYSQLUSER -p$MYSQLPASS -h"$MYSQLHOST1" -e"show status like '%tls%';"


❯ # SERVER 2
mysql -u$MYSQLUSER -p$MYSQLPASS -h"$MYSQLHOST2" -e"show variables like '%version%';"
+-------------------------+------------------------------------------------------+
| Variable_name           | Value                                                |
+-------------------------+------------------------------------------------------+
| innodb_version          | 5.6.40-84.0                                          |
| protocol_version        | 10                                                   |
| slave_type_conversions  |                                                      |
| tls_version             | TLSv1.1,TLSv1.2                                      |
| version                 | 5.6.40-84.0-log                                      |
| version_comment         | Percona Server (GPL), Release 84.0, Revision 47234b3 |
| version_compile_machine | x86_64                                               |
| version_compile_os      | Linux                                                |
| version_suffix          | -log                                                 |
+-------------------------+------------------------------------------------------+
mysql -u$MYSQLUSER -p$MYSQLPASS -h"$MYSQLHOST2" -e"show status like '%tls%';"
```

I've captured a test session on these servers here: 

https://asciinema.org/a/0lw8eG70VhihWcl6Al7i9MKXO?theme=solarized-light

Both of them say:
```
ssl.SSLError: [SSL: UNSUPPORTED_PROTOCOL] unsupported protocol (_ssl.c:1076)                                                  
```

**Comment by daveisfera** [comment]
I believe that @methane was correct, because I'm no longer seeing this issue now that I updated the database I was connecting to to use TLS 1.2

**Comment by MatrixManAtYrService** [comment]
I've worked around my issue by just using a different image. So I'm really only posting here in the spirit of leaving behind useful bits for others.

It does seem that updating the server is the *right* way, but I'd have to burn political capital to get that done.  In my case switching images was an easy thing to do, but I can imagine cases where that's troublesome too.

Ideally there would be a change we could make in the source container that add support for whatever version of TLS is involved (even if it was left out by whoever maintains that image).  I'm not knowledgeable enough about TLS or mysql to know what that change would be though.

**Comment by methane** [comment]
I said to [pass the customized SSL context](https://github.com/PyMySQL/PyMySQL/issues/817#issuecomment-546223292) [repeatedly](https://github.com/PyMySQL/PyMySQL/issues/817#issuecomment-549402078).  But no one try it.

Or you may be able to edit `/etc/ssl/openssl.conf` from 

```
[system_default_sect]
MinProtocol = TLSv1.2
```

to

```
[system_default_sect]
MinProtocol = TLSv1.1
```

But I'm not sure.  I am not expert of ssl.

**Comment by MatrixManAtYrService** [comment]
Thanks for the code snippet @methane, I tried passing the context explicitly, as you indicated, and got the same result.

I also found that making the modification to `/etc/ssl/openssl.conf` that you recommended, but still had the same problem.

------

I was doing some reading here: https://superuser.com/questions/1276703/ssl-unsupported-protocol-error and found this command: `openssl s_client -connect google.com:443 -tls1_2`

If I run it as-is (targeting google, with either tls1_1 or tls1_2) I get many details about the TLS configuration.  When I try it targeting the servers I'm having this problem with, I get this error:

```
140411349164480:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../ssl/record/ssl3_record.c:332:
```
That happens whether I'm on my local machine (where pymysql works), or whether I'm in the docker container (where it doesn't). My current theory about this is that my servers are misconfigured in some way, and there's something about the newer versions of openssl that is unwilling to talk to them due to that misconfiguration (where maybe the older versions were letting me get away with it).
